README.md
# cms.psps

<!-- badges: start -->
[![R-CMD-check](https://github.com/dewittpe/cms.psps/workflows/R-CMD-check/badge.svg)](https://github.com/dewittpe/cms.psps/actions)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/cms.psps)](https://cran.r-project.org/package=cms.psps)
<!-- badges: end -->

R package for downloading and installing the Physician/Supplier Procedure Summary Data provided by the Centers for Medicare and Medicaid


NEWS.md
# Version 0.1.0

**Initial Release**

## Features

* Automated download of the Physician/Suppler Procedure Summary (PSPS) data sets,
  calendar years 2010 through 2019, provided by the Centers for Medicare and
  Medicaid

* Simple import methods to read the PSPS data sets into R as data.tables.



R/data.R
#' Download the PSPS Data
#'
#' Download the public files provided by CMS.
#'
#' @param path directory to save the downloaded files too.
#' @param years years to download
#' @param ... not currently used.
#'
#' @export
psps_download <- function(path = NULL, years = 2010:2019, ...) {

  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  stopifnot(all(years %in% 2010:2019))

  # create needed directories, if needed
  dir.create(path = path, showWarnings = FALSE, recursive = TRUE)

  Map(f =
      function(n, u) {
        message(sprintf("Downloading %s:\n", n))
        utils::download.file(url = u, destfile = paste(path, n, sep = "/"))
      },
      n = names(psps_urls())[which(2010:2019 %in% years)],
      u = psps_urls()[which(2019:2019 %in% years)])
}

#' Verify PSPS .zip files via MD5SUM
#'
#' @param path directory to the downloaded PSPS files.
#' @param years data sets to verify
#' @param ... not currently used.
#'
#' @export
psps_md5sum <- function(path = NULL, years = 2010:2019, ...) {

  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  stopifnot(all(years %in% 2010:2019))

  # updated 6 May 2021 for v0.1.0
  expected <- list(
                     "psps2010.zip" = "ceef3a0f080eab632082340477a53d11"
                   , "psps2011.zip" = "a080b5f6ba4aafe350d5c03bb0298cab"
                   , "psps2012.zip" = "b6e6855d58d5cee5469a1c2c57c8f10b"
                   , "psps2013.zip" = "ebd6eff8594b474387b77bc025a305b5"
                   , "psps2014.zip" = "e5f7850e63dd1ef7d68f8c0cfb30f1ec"
                   , "psps2015.zip" = "ed85399041533aad50a6b0f441c2ac76"
                   , "psps2016.zip" = "8ea9b0656002b92bcd94226154f02d17"
                   , "psps2017.zip" = "114852c334e4a5c255f82ec48379748e"
                   , "psps2018.zip" = "5d3e55669ea9523250fe49ae2f41bf0f"
                   , "psps2019.zip" = "80ca784297c18c30acfe9edbf2e4be8d"
                   )
  expected <- expected[which(2010:2019 %in% years)]

  observed <- tools::md5sum(files = paste(path, names(expected), sep = "/"))
  names(observed) <- basename(names(observed))

  check <-
    merge(data.frame(file = names(observed), md5sum = observed,             row.names = NULL, stringsAsFactors = FALSE),
          data.frame(file = names(expected), md5sum = do.call(c, expected), row.names = NULL, stringsAsFactors = FALSE),
          by = "file",
          suffixes = c(".observed", ".expected"))
  check$check <- sapply(check$md5sum.observed == check$md5sum.expected, isTRUE)
  check
}

#' Unzip Downloaded Data Sets
#'
#' @param path directory to the downloaded PSPS files.
#' @param ... not currently used.
#'
#' @export
psps_unzip <- function(path = NULL, ...) {

  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  zips <- list.files(path, pattern = "*\\.zip$", full.names = TRUE)
  zips <- stats::setNames(zips, basename(list.files(path, pattern = "*\\.zip$", full.names = TRUE)))

  for (i in seq_along(zips)) {
    message(paste0("Unzipping ", names(zips)[i]))
    utils::unzip(zips[i],
                 exdir = paste0(path, "/", sub(".zip", "", names(zips)[i])))
  }
  invisible(TRUE)
}

psps_urls <- function() {
  # urls aquired from https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/Physician-Supplier-Procedure-Summary/index.html

  # v0.1.0
  list(
         "psps2010.zip" = "https://downloads.cms.gov/files/PSPS2010.zip"
       , "psps2011.zip" = "https://www.cms.gov/files/zip/psps2011suppress.zip"
       , "psps2012.zip" = "https://www.cms.gov/files/zip/psps2012suppress.zip"
       , "psps2013.zip" = "https://www.cms.gov/files/zip/psps2013suppress.zip"
       , "psps2014.zip" = "https://www.cms.gov/files/zip/psps2014suppress.zip"
       , "psps2015.zip" = "https://www.cms.gov/files/zip/2015-physician-supplier-procedure-summary.zip"
       , "psps2016.zip" = "https://www.cms.gov/files/zip/2016-physician-supplier-procedure-summary.zip"
       , "psps2017.zip" = "https://www.cms.gov/files/zip/2017-physician-supplier-procedure-summary.zip"
       , "psps2018.zip" = "https://www.cms.gov/files/zip/2018-physician-supplier-procedure-summary.zip"
       , "psps2019.zip" = "https://www.cms.gov/files/zip/2019-physician-supplier-procedure-summary.zip"
       )
}











R/import.R
#' Import Functions for the PSPS data
#'
#' @param path path to the directory where the downloaded data is stored.
#' @param years which years to import
#' @param ... additional arguments passed to \code{data.table::fread}
#'
#' @return A \code{\link[data.table]{data.table}}
#'
#' @examples
#' \dontrun{
#'
#' library(cms.psps)
#'
#' # Check that the data is as expected
#' check <- psps_md5sum()
#' if (!all(check$check)) {
#'   psps_download()
#'   psps_unzip()
#' }
#'
#' psps_2010 <- psps_import(year = 2010)
#' psps_2010
#'
#' psps_2018_2019 <- psps_import(years = 2018:2019)
#' psps_2018_2019
#'
#' }
#'
#' @name import
NULL

#' @rdname import
#' @export
psps_import <- function(path = NULL, years = 2010:2019, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  DT <-
    lapply(years,
           function(yr) {
               data.table::fread(
                                 file = paste0(path, "/psps", yr, "/PSPS_", yr, "_SUPPRESS.csv")
                                 , colClasses = c(rep(character(), 8), rep(numeric(), 7), rep(character(), 3))
             )
           })

  DT <- stats::setNames(DT, years)
  DT <- data.table::rbindlist(DT, idcol = "YEAR")
  data.table::set(DT, j = "YEAR", value = as.integer(DT[["YEAR"]]))
  DT
}



vignette-spinners/cms.psps-introduction.R
#'---
#'title: "Introduction to the cms.PSPS Package"
#'author: Peter DeWitt
#'output: rmarkdown::html_vignette
#'---
#'
#+ label = "setup", include = FALSE
knitr::opts_chunk$set(collapse = TRUE)

#'
#' The purpose of this package is to provide quick and easy coding for
#' downloading and importing the
#' (Physician/Suppler Procedure Summary)[https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/Physician-Supplier-Procedure-Summary]
#' data provided by the Centers for Medicare and Medicaid.
#'
#'>The Physician/Supplier Procedure Summary (PSPS) file is a summary of calendar year Medicare Part B carrier and durable medical equipment fee-for-service claims. The file is organized by carrier, pricing locality, Healthcare Common Procedure Coding System (HCPCS) code, HCPCS modifier, provider specialty, type of service, and place of service. The summarized fields are total submitted services and charges, total allowed services and charges, total denied services and charges, and total payment amounts.
#'>
#'>Note: As of November 2020, the following suppression logic will apply:
#'>
#'> Submitted Service Count and Submitted Charge Amount will be suppressed for rows where Submitted Service Count is less than 11; Denied Service Count and Denied Charge Amount will be suppressed for rows where Denied Service Count is less than 11; Assigned Service Count and Allowed Charge Amount will be suppressed for rows where Assigned Service Count is less than 11
#'
#' As will be shown in the following, the size of the data sets is non-trivial.
#' It is recommended that you load and use the data.table package to work with
#' the data.  The import functions provide the data as data.tables.
#'
#+ label = "namespaces"
# /*
if (interactive()) {
  devtools::load_all()
} else {
# */
library(cms.psps)
# /*
}
# */
library(data.table)

#'
#' # Initial Steps
#'
#' Two steps are needed, download and unzip.  Below are some details.  A
#' function for checking the md5sums of the download files has been provided
#' too.
#'
#' ## Downloading PSPS
#'
#' For version 0.1.0 of the package there are tools for downloading and
#' importing the PSPS data from years 2010 through 2019.  The data files are
#' large.  On disk size is not bad, but the size in RAM could be restrictive.
#'
#' You can download the PSPS as .zip file from CMS by calling psps_download().
#' The default is to download all the data sets, but you can specify the years
#' you want.  By default the data is stored on your system in the location
#' defined by
#'
rappdirs::user_data_dir(appname = "cms.psps")
#'
#'
#+ label = "psps_download", eval = FALSE
# /*
while (FALSE) {
# */
psps_download()
# /*
}
# */

#'
#' ## Checking download
#'
#' Calling psps_md5sum will check the md5sum of the downloaded .zip files.
#'
psps_md5sum()

#'
#' ## Unziping the Downloaded Files
#'
#' Simply calling psps_will unzip the files.
#+ eval = FALSE
# /*
while (FALSE) {
# */
psps_unzip()
# /*
}
# */

#'
#' # Importing Data
#'
#' To start to work with the PSPS data you need only call psps_import().  The
#' import is as if you read in the csv with one additional column YEAR.
#'
#+ label = "import 2017 data"
psps2017 <- psps_import(years = 2017)
str(psps2017)

#'
#' There are
{{ formatC(nrow(psps2017), format = "d", big.mark = ",") }}
#' rows in that file.  The approximate size in memory is:
#'
format(utils::object.size(psps2017), units = "GB")

#'
#' Reading in multiple years will result in one data.table.
psps_14_15 <- psps_import(years = 2014:2015)
str(psps_14_15)

psps_14_15[, formatC(.N, format = "d", big.mark = ","), by = .(YEAR)]

format(utils::object.size(psps_14_15), units = "GB")

#'
#' # Data Set Contents
#'
#' This information has been copied from the CMS provided documentation.
#'
#' ## Column 1 : YEAR
#' This was added by the import steps to denote the year the data come from
#'
#' ## Column 2 : HCPCS_CD
#'
#' HCPCS Code -
#'
#' The Health Care Common Procedure Coding System (HCPCS) is a collection of
#' codes that represent procedures, supplies, products and services which may be
#' provided to Medicare beneficiaries and to individuals enrolled in private
#' health insurance programs.  The codes are divided into three levels, or
#' groups as described below:
#'
#' Level I: Codes and descriptors copyrighted by the American Medical
#' Association's Current Procedural Terminology, Fourth Edition (CPT-4).  These
#' are 5 position numeric codes representing physician and nonphysician
#' services.
#'
#' CPT-4 codes including both long and short descriptions shall be used in
#' accordance with the CMS/AMA agreement.  Any other use violates the AMA
#' copyright.
#'
#' Level II Includes codes and descriptors copyrighted by the American Dental
#' Association's Current Dental Terminology, Third Edition (CDT-3).  These are 5
#' position alpha-numeric codes comprising the D series.  All other level II
#' codes and descriptors are approved and maintained jointly by the
#' alpha-numeric editorial panel (consisting of CMS, the Health Insurance
#' Association of America, and the Blue Cross and Blue Shield Association).
#' These are 5 position alpha- numeric codes representing primarily items and
#' nonphysician services that are not represented in the level I codes.
#'
#' Level III Codes and descriptors developed by Medicare carriers for use at the
#' local (carrier) level.  These are 5 position alpha-numeric codes in the W, X,
#' Y or Z series representing physician and nonphysician services that are not
#' represented in the level I or level II codes.
#'
#' ## Column 3 : HCPCS_INITIAL_MODIFIER_CD
#'
#' HCPCS Initial Modifier Code -
#'
#' A first modifier to the HCPCS procedure code to enable a more specific
#' procedure identification for the line item service on the noninstitutional
#' claim.
#'
#' ## Column 4: PROVIDER_SPEC_CD
#'
#' Provider Specialty Code -
#'
#' CMS specialty code used for pricing the line item service on the
#' noninstitutional claim.
#'
#' ## Column 5: CARRIER_NUM
#'
#' Carrier Number -
#'
#' The identification number assigned by CMS to a carrier authorized to process
#' claims from a physician or supplier.
#'
#' ## Column 6: PRICING_LOCALITY_CD
#'
#' Pricing Locality Code -
#'
#' Code denoting the carrier-specific locality used for pricing the service for
#' this line item on the carrier claim (non-DMERC).  For DMERCs, this field
#' contains the beneficiary SSA State Code
#'
#' ## Column 7: TYPE_OF_SERVICE_CD
#'
#' Type of Service Code -
#'
#' Code indicating the type of service, as defined in the CMS Medicare Carrier
#' Manual, for this line item on the non-institutional claim.
#'
#' ## Column 8: PLACE_OF_SERVICE_CD
#'
#' Place of Service Code -
#'
#' The code indicating the place of service, as defined in the Medicare Carrier
#' Manual, for this line item on the noninstitutional claim.
#'
#' ## Column 9: HCPCS_SECOND_MODIFIER_CD
#'
#' HCPCS Second Modifier Code -
#'
#' A second modifier to the HCPCS procedure code to make it more specific than
#' the first modifier code to identify the line item procedures for this claim.
#'
#' ## Column 10: SUBMITTED_SERVICE_CNT
#'
#' Physician/Supplier Procedure Summary (PSPS) Submitted Service Count  -
#'
#' The count of the total number of submitted services.
#'
#' ## Column 11: SUBMITTED_CHARGE_AMT
#'
#' Physician/Supplier Procedure Summary (PSPS) Submitted Charge Amount
#'
#' The amount of charges submitted by the provider to Medicare.
#'
#' ## Column 12: ALLOWED_CHARGE_AMT
#'
#' Physician/Supplier Procedure Summary (PSPS) Allowed Charge Amount -
#'
#' The amount that is approved (allowed) for Medicare.
#'
#' ## Column 13: DENIED_SERVICES_CNT
#'
#' Physician/Supplier Procedure Summary (PSPS) Denied Services Count -
#'
#' The count of the number of submitted services that are denied by Medicare.
#'
#' ## Column 14: DENIED_CHARGE_AMT
#'
#' Physician/Supplier Procedure Summary (PSPS) Denied Charge Amount -
#'
#' The amount of submitted charges for which Medicare payment was denied.
#'
#' ## Column 15: ASSIGNED_SERVICES_CNT
#'
#' Physician/Supplier Procedure Summary (PSPS) Assigned Services Count -
#'
#' The count of the number of services from providers accepting Medicare assignment.
#'
#' ## Column 16: NCH_PAYMENT_AMT
#'
#' Physician/Supplier Procedure Summary (PSPS) NCH Payment Amount -
#'
#' The amount of payment made from the trust fund (after deductible and
#' coinsurance amounts have been paid).
#'
#' ## Column 17: HCPCS_ASC_IND_CD
#'
#' Physician/Supplier Procedure Summary (PSPS) HCPCS ASC Indicator Code -
#'
#' A Y/N code used to indicate whether the procedure is approved to be performed
#' in an Ambulatory Surgical Center (ASC).
#'
#' ## Column 18: ERROR_IND_CD
#'
#' Physician Supplier Procedure Summary (PSPS) Error Indicator Code -
#'
#' The code used to indicate combinations of errors on key fields.
#'
#' ## Column 19: BETOS_CD
#'
#' HCPCS Berenson-Eggers Type of Service Code (BETOS) -
#'
#' This field is valid beginning with 2003 data.  The Berenson-Eggers Type of
#' Service (BETOS) for the procedure code based on generally agreed upon
#' clinically meaningful groupings of procedures and services.



R/spark_import.R
#' Spark Import Functions for the PSPS data
#'
#' Importing the PSPS data into spark via sparklyr
#'
#' @param sc a sparklyr cluster
#' @param path path to the directory where the downloaded data is stored.
#' @param envir the envir to assign the \code{tbl_spark} object too, defatuls to
#' the global environment.
#' @param ... arguments passed to \code{\link[sparklyr]{spark_read_csv}} (for
#' the import of 2017 data) or to \code{\link[sparklyr]{spark_read_text}} for
#' the years 2004 through 2016.
#'
#' @return A \code{tbl_spark} object of the name psps_<year> will be in the
#' \code{envir}.
#'
#' @examples
#' \donttest{
#'
#' library(cms.psps)
#'
#' # Check that the data is as expected
#' check <- psps_md5sum()
#' if (!all(check$check)) {
#'   psps_download()
#'   psps_unzip()
#' }
#'
#' # Set up spark if needed
#' # sparklyr::spark_install()
#'
#' sc <- sparklyr::spark_connect("local")
#'
#' ls()
#'
#' spark_psps_import_2010(sc, memory = FALSE)
#' psps_2010
#'
#' spark_psps_import_2011(sc, memory = FALSE)
#' psps_2011
#'
#' spark_psps_import_2012(sc, memory = FALSE)
#' psps_2012
#'
#' spark_psps_import_2013(sc, memory = FALSE)
#' psps_2013
#'
#' spark_psps_import_2014(sc, memory = FALSE)
#' psps_2014
#'
#' spark_psps_import_2015(sc, memory = FALSE)
#' psps_2015
#'
#' # Be aware: There are some non utf8 characters in the data.  Be aware and
#' # explore and test your data exports.  In the example below there is a
#' library(magrittr)
#' psps_2015 %>%
#'   dplyr::group_by(HCPCS_INITIAL_MODIFIER_CD) %>%
#'   dplyr::summarize(n()) %>%
#'   dplyr::arrange(HCPCS_INITIAL_MODIFIER_CD) %>%
#'   dplyr::collect() %>%
#'   dplyr::filter(grepl("^G.*", HCPCS_INITIAL_MODIFIER_CD)) %>%
#'   print(n = Inf)
#'
#' spark_psps_import_2016(sc, memory = FALSE)
#' psps_2016
#'
#' spark_psps_import_2017(sc, memory = FALSE)
#' psps_2017
#'
#' spark_psps_import_2018(sc, memory = FALSE)
#' psps_2018
#'
#' ls()
#'
#' sparklyr::spark_disconnect(sc)
#'
#' }
#'
#'
#' @name spark_import
NULL

#' @rdname spark_import
#' @export
spark_psps_import_2010 <- function(sc, path = NULL, envir = .GlobalEnv, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  files <- list.files(path = paste(path, "psps2010", sep = "/"),
                      full.names = TRUE,
                      pattern = "PSPS\\d{2}")

  object <-
    sparklyr::sdf_bind_rows(lapply(files, function(x, ...) {
                                     sparklyr::spark_read_text(sc, name = x, path = x, ...)
                                   }))

  assign(x     = "psps_2010",
         value = psps_mutate_v2(object, YEAR = 2010L),
         envir = envir)

  invisible(TRUE)
}

#' @rdname spark_import
#' @export
spark_psps_import_2011 <- function(sc, path = NULL, envir = .GlobalEnv, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  files <- list.files(path = paste(path, "psps2011", sep = "/"),
                      full.names = TRUE,
                      pattern = "PSPS\\d{2}")

  object <-
    sparklyr::sdf_bind_rows(lapply(files, function(x, ...) {
                                     sparklyr::spark_read_text(sc, name = x, path = x, ...)
                                   }))

  assign(x     = "psps_2011",
         value = psps_mutate_v2(object, YEAR = 2011L),
         envir = envir)

  invisible(TRUE)
}

#' @rdname spark_import
#' @export
spark_psps_import_2012 <- function(sc, path = NULL, envir = .GlobalEnv, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  files <- list.files(path = paste(path, "psps2012", sep = "/"),
                      full.names = TRUE,
                      pattern = "PSPS\\d{2}")

  object <-
    sparklyr::sdf_bind_rows(lapply(files, function(x, ...) {
                                     sparklyr::spark_read_text(sc, name = x, path = x, ...)
                                   }))

  assign(x     = "psps_2012",
         value = psps_mutate_v2(object, YEAR = 2012L),
         envir = envir)

  invisible(TRUE)
}

#' @rdname spark_import
#' @export
spark_psps_import_2013 <- function(sc, path = NULL, envir = .GlobalEnv, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  files <- list.files(path = paste(path, "psps2013", sep = "/"),
                      full.names = TRUE,
                      pattern = "PSPS\\d{2}\\.(txt|TXT)$")

  object <-
    sparklyr::sdf_bind_rows(lapply(files, function(x, ...) {
                                     sparklyr::spark_read_text(sc, name = x, path = x, ...)
                                   }))

  assign(x     = "psps_2013",
         value = psps_mutate_v2(object, YEAR = 2013L),
         envir = envir)

  invisible(TRUE)
}

#' @rdname spark_import
#' @export
spark_psps_import_2014 <- function(sc, path = NULL, envir = .GlobalEnv, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  files <- list.files(path = paste(path, "psps2014", sep = "/"),
                      full.names = TRUE,
                      pattern = "PSPS\\d{2}\\.(txt|TXT)$")

  object <-
    sparklyr::sdf_bind_rows(lapply(files, function(x, ...) {
                                     sparklyr::spark_read_text(sc, name = x, path = x, ...)
                                   }))

  assign(x     = "psps_2014",
         value = psps_mutate_v2(object, YEAR = 2014L),
         envir = envir)

  invisible(TRUE)
}

#' @rdname spark_import
#' @export
spark_psps_import_2015 <- function(sc, path = NULL, envir = .GlobalEnv, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  files <- list.files(path = paste(path, "psps2015", sep = "/"),
                      full.names = TRUE,
                      pattern = "PSPS\\d{2}\\.(txt|TXT)$")

  object <-
    sparklyr::sdf_bind_rows(lapply(files, function(x, ...) {
                                     sparklyr::spark_read_text(sc, name = x, path = x, ...)
                                   }))

  assign(x     = "psps_2015",
         value = psps_mutate_v2(object, YEAR = 2015L),
         envir = envir)

  invisible(TRUE)
}

#' @rdname spark_import
#' @export
spark_psps_import_2016 <- function(sc, path = NULL, envir = .GlobalEnv, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  files <- list.files(path = paste(path, "psps2016", sep = "/"),
                      full.names = TRUE,
                      pattern = "PSPS\\d{2}\\.(txt|TXT)$")

  object <-
    sparklyr::sdf_bind_rows(lapply(files, function(x, ...) {
                                     sparklyr::spark_read_text(sc, name = x, path = x, ...)
                                   }))

  assign(x     = "psps_2016",
         value = psps_mutate_v2(object, YEAR = 2016L),
         envir = envir)

  invisible(TRUE)
}

#' @rdname spark_import
#' @export
spark_psps_import_2017 <- function(sc, path = NULL, envir = .GlobalEnv, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  assign(x     = "psps_2017",
         value =
           dplyr::mutate(sparklyr::spark_read_csv(sc = sc,
                                          name = "psps_2017",
                                          path = paste(path, "psps2017", "PSPS_2017.csv", sep = "/"),
                                          colClasses = psps_colClasses,
                                          ...),
                         YEAR = 2017L),
         envir = envir)

  invisible(TRUE)
}

#' @rdname spark_import
#' @export
spark_psps_import_2018 <- function(sc, path = NULL, envir = .GlobalEnv, ...) {
  if (is.null(path)) {
    path <- rappdirs::user_data_dir(appname = "cms.psps")
  }

  assign(x     = "psps_2018",
         value =
           dplyr::mutate(sparklyr::spark_read_csv(sc = sc,
                                          name = "psps_2018",
                                          path = paste(path, "psps2018", "PSPS2018.csv", sep = "/"),
                                          colClasses = psps_colClasses,
                                          ...),
                         YEAR = 2018L),
         envir = envir)

  invisible(TRUE)
}


# colClasses
psps_colClasses <-
  c(
    "HCPCS_CD"                   = "character",
    "HCPCS_INITIAL_MODIFIER_CD"  = "character",
    "PROVIDER_SPEC_CD"           = "character",
    "CARRIER_NUM"                = "character",
    "PRICING_LOCALITY_CD"        = "character",
    "TYPE_OF_SERVICE_CD"         = "character",
    "PLACE_OF_SERVICE_CD"        = "character",
    "HCPCS_SECOND_MODIFIER_CD"   = "character",
    "REGION"                     = "character",
    "PSPS_SUBMITTED_SERVICE_CNT" = "numeric",
    "MTUS"                       = "character",
    "ANESTHESIA_UNITS"           = "character",
    "MTUIND"                     = "character",
    "PSPS_SUBMITTED_CHARGE_AMT"  = "numeric",
    "PSPS_ALLOWED_CHARGE_AMT"    = "numeric",
    "PSPS_DENIED_SERVICES_CNT"   = "numeric",
    "PSPS_DENIED_CHARGE_AMT"     = "numeric",
    "PSPS_ASSIGNED_SERVICES_CNT" = "numeric",
    "PSPS_NCH_PAYMENT_AMT"       = "numeric",
    "PSPS_HCPCS_ASC_IND_CD"      = "character",
    "PSPS_ERROR_IND_CD"          = "character",
    "HCPCS_BETOS_CD"             = "character"
    )

psps_mutate_v2 <- function(.data, ...) {
  # This is based on the PSPS File Desc Layout as of 08/29/2010
  # .data <- sparklyr::spark_apply(.data, function(x) {x$line <- iconv(x$line, from = "Latin1", to = "UTF8"); x})
  dplyr::mutate(.data,
                HCPCS_CD                   =            substr(.data$line, start =   1L, stop =   5L),  #1
                HCPCS_INITIAL_MODIFIER_CD  =            substr(.data$line, start =   6L, stop =   7L),  #2
                PROVIDER_SPEC_CD           =            substr(.data$line, start =   8L, stop =   9L),  #3
                CARRIER_NUM                =            substr(.data$line, start =  10L, stop =  14L),  #4
                PRICING_LOCALITY_CD        =            substr(.data$line, start =  15L, stop =  16L),  #5
                TYPE_OF_SERVICE_CD         =            substr(.data$line, start =  17L, stop =  17L),  #6
                PLACE_OF_SERVICE_CD        =            substr(.data$line, start =  18L, stop =  19L),  #7
                HCPCS_SECOND_MODIFIER_CD   =            substr(.data$line, start =  20L, stop =  21L),  #8
                PSPS_SUBMITTED_SERVICE_CNT = as.numeric(substr(.data$line, start =  22L, stop =  35L)), #9
                PSPS_SUBMITTED_CHARGE_AMT  = as.numeric(substr(.data$line, start =  36L, stop =  48L)), #10
                PSPS_ALLOWED_CHARGE_AMT    = as.numeric(substr(.data$line, start =  49L, stop =  61L)), #11
                PSPS_DENIED_SERVICES_CNT   = as.numeric(substr(.data$line, start =  62L, stop =  75L)), #12
                PSPS_DENIED_CHARGE_AMT     = as.numeric(substr(.data$line, start =  76L, stop =  88L)), #13
                PSPS_ASSIGNED_SERVICES_CNT = as.numeric(substr(.data$line, start =  89L, stop = 102L)), #14
                PSPS_NCH_PAYMENT_AMT       = as.numeric(substr(.data$line, start = 103L, stop = 115L)), #15
                PSPS_HCPCS_ASC_IND_CD      =            substr(.data$line, start = 116L, stop = 116L),  #16
                PSPS_ERROR_IND_CD          =            substr(.data$line, start = 117L, stop = 118L),  #17
                HCPCS_BETOS_CD_CD          =            substr(.data$line, start = 119L, stop = 121L),  #18
                `line`                     = NULL,
                ...
                )
}

psps_mutate_v1 <- function(.data, ...) {
  # this is good for years 2004 - 2008 (not public)
  # .data$`line` <- iconv(.data$line, from = "Latin1", to = "UTF8")
  dplyr::mutate(.data,
                HCPCS_CD                    =            substr(.data$line, start =   1L, stop =   5L),
                HCPCS_INITIAL_MODIFIER_CD   =            substr(.data$line, start =   6L, stop =   7L),
                PROVIDER_SPEC_CD            =            substr(.data$line, start =   8L, stop =   9L),
                CARRIER_NUM                 =            substr(.data$line, start =  10L, stop =  14L),
                PRICING_LOCALITY_CD         =            substr(.data$line, start =  15L, stop =  16L),
                TYPE_OF_SERVICE_CD          =            substr(.data$line, start =  17L, stop =  17L),
                PLACE_OF_SERVICE_CD         =            substr(.data$line, start =  18L, stop =  19L),
                HCPCS_SECOND_MODIFIER_CD    =            substr(.data$line, start =  20L, stop =  21L),
                REGION                      =            substr(.data$line, start =  22L, stop =  23L),
                PSPS_SUBMITTED_SERVICE_CNT  = as.numeric(substr(.data$line, start =  24L, stop =  32L)),
                MTUS                        = as.numeric(substr(.data$line, start =  33L, stop =  41L)),
                ANESTHESIA_UNITS            = as.numeric(substr(.data$line, start =  33L, stop =  41L)),
                MTUIND                      =            substr(.data$line, start =  42L, stop =  42L),
                PSPS_SUBMITTED_CHARGE_AMT   = as.numeric(substr(.data$line, start =  43L, stop =  52L)),
                PSPS_ALLOWED_CHARGE_AMT     = as.numeric(substr(.data$line, start =  53L, stop =  62L)),
                PSPS_DENIED_SERVICES_CNT    = as.numeric(substr(.data$line, start =  63L, stop =  70L)),
                PSPS_DENIED_CHARGE_AMT      = as.numeric(substr(.data$line, start =  71L, stop =  78L)),
                PSPS_ASSIGNED_SERVICES_CNT  = as.numeric(substr(.data$line, start =  79L, stop =  87L)),
                PSPS_NCH_PAYMENT_AMT        = as.numeric(substr(.data$line, start =  88L, stop =  97L)),
                PSPS_HCPCS_ASC_IND_CD       =            substr(.data$line, start =  98L, stop =  98L),
                PSPS_ERROR_IND_CD           =            substr(.data$line, start =  99L, stop =  99L),
                HCPCS_BETOS_CD              =            substr(.data$line, start = 100L, stop = 105L),
                `line`                      = NULL,
                ...
  )
}

